### ISSUE TITLE
Aliases fail with some providers

### ISSUE THREAD

**Issue by hongkongkiwi** [hashicorp/terraform#25902]
When using the hashicorp Docker provider, terraform does not correctly handle aliases.

```
Terraform v0.13.0
+ provider registry.terraform.io/terraform-providers/docker v2.7.2
```

For example, this will fail with error during terraform validate:
`Error: missing provider provider["registry.terraform.io/hashicorp/docker"].foo`

```
provider "docker" {
  alias = "foo"
  host = "tcp://127.0.0.1:2376/"
}

module "mycustommodule" {
  source = "./test2"
  providers = {
    docker = docker.foo
  }
}
```

**Comment by wenerme** [comment]
Any workaround ? blocked by this

**Comment by hongkongkiwi** [comment]
I've been using the provider config inside the module. This allows me construct the host param using variables passed to the module.

BUT, the only downside of this is you cannot use the for_each or count feature introduced in 0.13. That's where aliases make a bit more sense as using these features won't allow you to embed the provider config inside the module.

**Comment by mildwonkey** [comment]
hi @hongkongkiwi ! 

To pass an aliased provider into a module, the module itself needs to declare the same provider and alias, as documented [here](https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly).

Given your example, the module "mycustommodule" will need to have this configuration somewhere:

```
provider "docker" {
  alias = "foo"
}
```

If "mycustommodule" does contain this block and is giving you this error, please open a new issue and fully fill out the issue template, including a *full* configuration (including the module) that reproduces the issue, as well as the `TRACE` logs from the failing terraform command.

**Comment by hfrog** [comment]
Hello.
I've stuck with the same error too.
my config is:
`terraform {
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "~> 3.0"
    }
    google-beta = {
      source  = "hashicorp/google-beta"
      version = "~> 2.0"
    }
  }
  required_version = ">= 0.13"
}

provider "google-beta" {
  region  = "europe-west1"
  alias = "legacy"
}

provider "google" {
  region  = "europe-west1"
}

module "custom_module" {
  source = "./custom_module"

  providers = {
    google = google-beta.legacy
  }

}`

Debug output of 'terraform plan' is below:
https://gist.github.com/hfrog/4400f6fba7490d7a7fb4669bc6f7ce45

What is the correct way to configure aliases in terraform 0.13 ?
The block
`  providers = {
    google = google-beta.legacy
  }
`
worked perfect in terraform 0.12.

**Comment by mildwonkey** [comment]
@hfrog that's a different issue, and unfortunately not something that we had deliberately supported in terraform v0.12. 

In v0.13, the provider source address [is a _global_ setting](https://www.terraform.io/docs/configuration/modules.html#provider-version-constraints-in-modules). You cannot use the `hashicorp/google-beta` provider in a module that declares that the `google` provider's source is `"hashicorp/google"`. 

Since you're describing something that's different from the original issue, I'd suggest you use [the community forum](https://discuss.hashicorp.com/c/terraform-core) where there are far more people ready to help, whereas the GitHub issues here are generally monitored only by our few core maintainers, or open a new github issue.
